src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java

```diff
@@ -318,10 +322,31 @@ private void runActiveTurn(Game currentGame, Integer cardId) throws InterruptedE
     }
 
     public void handleInactivePlayer(Integer gameId) {
-        if (isGameInCalc(gameId)) return;
+        if (isGameInCalc(gameId)) {
+            deferredExecutionService.deferTask(gameId, () -> executeInactivePlayerLogic(gameId));
+        } else {
+            executeInactivePlayerLogic(gameId);
+        }
+    }
+
+    @EventListener
+    public void onGameCalculationStatusChange(InactiveRequestEvent event) {
+        if (deferredExecutionService.hasDeferredTask(event.getGameId())) {
+            deferredExecutionService.executeDeferredTask(event.getGameId());
+        }
+    }
+
+    private void executeInactivePlayerLogic(Integer gameId) {
         User inactivePlayer = UserContextHolder.getCurrentUser();
         Game currentGame = inMemoryGameRepository.findById(gameId);
-        if (currentGame != null && inactivePlayer.getUserId().equals(currentGame.getActivePlayer())) {
+
+        int historySize = currentGame.getHistory().size();
+
+        // We need to ensure that a correct selection at the very end of a turn does not in a wrongful change of Turn
+        if (historySize > 1 &&
+                (currentGame.getHistory().get(historySize - 1).getPicks().isEmpty()
+                        && currentGame.getHistory().get(historySize - 2).getUserId() == currentGame.getActivePlayer())) {return;}
```

Comment:
here calculates current turn is empty & last turn is of current activePlayer
Is it to check, if the user is on streak, then does not end turn?

Reply:
"here calculates current turn is empty & last turn is of current activePlayer" that is correct. The reason for it however is slightly different. We need these tests to check what happened before the inactive request was queued.

How it is implemented now, when the inactive request comes in we block the request if the game is calculating. Instead of discarding the req, we store/defer it in the DeferredExecService. So after a game is done calculating, we check if in the meantime we have stored an inactive request. (Note: we only use this for the inactive request, hence it is only possible for one request to be stored, and that is all that we require for this case).

Case 0: Turn User 1, Picks Card 2 out of 2, Sends Inactive Request (which we store), Card 2 does not match. That means in the meantime, since the picks do not match, the active player has been updated (not equal to inactive request maker) hence we do not process the inactive request -> not caught by any if stmt. 

Case 1 (2nd if stmt): Turn User 1, Picks card 1 out of 2, Sends Inactive Request (which we store). That means that the game updates, which blocks the request, however, once the calculation is over, we need to process the inactive request since the player has not matched a set in their turn. We do that by simply checking if the active user is still the same as the one who made the request

Case 2 (1st if stmt): Turn User 1, Picks Card 2 out of 2, Sends Inactive Request (which we store), Card 2 does match. This case is a bit more strange to catch. Essentially, the player has completed their turn in time, but has not received the update that that was the case in the FE in time, and hence has sent an inactive request. However, we must disregard this inactive request, we do that by looking into the history. The only possibility, that this has happened, is if the game has ended the last turn (with user id 1) and has already created a new turn (also user 1). If that has happened, we know that the players choice was indeed in time and correct, hence we return without the inactive request.

Hope that helps